lib/tree.rb

```diff
@@ -1,62 +1,238 @@
+    inorder_helper(current_node.right, list)
+
+    return list
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def preorder_helper(current_node, list)
+    return list if current_node.nil?
+
+    # middle node
+    list << { key: current_node.key, value: current_node.value }
+
+    # left side
+    preorder_helper(current_node.left, list)
+
+    # right side
+    preorder_helper(current_node.right, list)
+
+    return list
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def postorder_helper(current_node, list)
+    return list if current_node.nil?
+
+    # left side
+    postorder_helper(current_node.left, list)
+
+    # right side
+    postorder_helper(current_node.right, list)
+
+    # middle node
+    list << { key: current_node.key, value: current_node.value }
+
+    return list
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(h) - h is height of tree
+  def height_helper(current_node)
+    if current_node.nil?
+      return 0
+    else
+      left_height = height_helper(current_node.left)
+      right_height = height_helper(current_node.right)
+
+      if left_height > right_height
+        return left_height + 1
+      else
+        return right_height + 1
+      end
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def height
-    raise NotImplementedError
+    return height_helper(@root)
   end
 
-  # Optional Method
-  # Time Complexity: 
-  # Space Complexity: 
+  # Optional Methods
+  # Time Complexity: O(n)
+  # Space Complexity: O(w) - w is width of tree
   def bfs
-    raise NotImplementedError
+    result = []
+
+    return result if @root.nil?
+
+    queue = [@root]
+
+    while !queue.empty?
+      node = queue.shift
+
+      result << { key: node.key, value: node.value }
+
+      if node.left
+        queue << node.left
+      end
+
+      if node.right
+        queue << node.right
+      end
+    end
+
+    return result
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def find_parent(current_node, key)
+    return nil if current_node.nil?
+
+    if current_node.left && (current_node.left.key == key) || current_node.right && (current_node.right.key == key)
+      return current_node
+    end
+
+    if key < current_node.key
+      current_node = find_parent(current_node.left, key)
+    else
+      current_node = find_parent(current_node.right, key)
+    end
+  end
+
+  def find_successor(current_node)
+    return current_node if current_node.right.nil?
+
+    find_successor(current_node.right)
+  end
+
+  def delete(key)
```

Comment:
This works, it could use some refactoring, but it does work.  Given what you have here I think the space/time complexity is O(log n) if the tree is balanced.